# Exploratory Data Analysis

## Import event data

In [1]:
import numpy as np
import pandas as pd

#Load all files
events = pd.read_csv('Inputs\\stage\\02-14-2018-staged.csv')
events = pd.concat([events, pd.read_csv('Inputs\\stage\\02-15-2018-staged.csv')], axis=0, ignore_index=True, copy=False)
events = pd.concat([events, pd.read_csv('Inputs\\stage\\02-16-2018-staged.csv')], axis=0, ignore_index=True, copy=False)
events = pd.concat([events, pd.read_csv('Inputs\\stage\\02-20-2018-staged.csv')], axis=0, ignore_index=True, copy=False)
events = pd.concat([events, pd.read_csv('Inputs\\stage\\02-21-2018-staged.csv')], axis=0, ignore_index=True, copy=False)
events = pd.concat([events, pd.read_csv('Inputs\\stage\\02-22-2018-staged.csv')], axis=0, ignore_index=True, copy=False)
events = pd.concat([events, pd.read_csv('Inputs\\stage\\02-23-2018-staged.csv')], axis=0, ignore_index=True, copy=False)
events = pd.concat([events, pd.read_csv('Inputs\\stage\\02-28-2018-staged.csv')], axis=0, ignore_index=True, copy=False)
events = pd.concat([events, pd.read_csv('Inputs\\stage\\03-01-2018-staged.csv')], axis=0, ignore_index=True, copy=False)
events = pd.concat([events, pd.read_csv('Inputs\\stage\\03-02-2018-staged.csv')], axis=0, ignore_index=True, copy=False)

print(events.shape)

(6229789, 79)


In [2]:
events

Dst Port  Protocol  Flow Duration  Tot Fwd Pkts  Tot Bwd Pkts  \
0               0         0      112641719             3             0   
1              22         6        6453966            15            10   
2              22         6        8804066            14            11   
3              22         6        6989341            16            12   
4              80         6         474670             5             3   
...           ...       ...            ...           ...           ...   
6229784      3389         6        4220892            14             8   
6229785      3389         6        4305082            14             8   
6229786        80         6        5598105             4             4   
6229787        53        17          35158             2             2   
6229788      3389         6        3998435            14             8   

         TotLen Fwd Pkts  TotLen Bwd Pkts  Fwd Pkt Len Max  Fwd Pkt Len Min  \
0                    0.0              0.0              0.0              0.0   
1                 1239.0           2273.0            744.0              0.0   
2                 1143.0           2209.0            744.0              0.0   
3                 1239.0           2273.0            744.0              0.0   
4                  214.0            466.0            214.0              0.0   
...                  ...              ...              ...              ...   
6229784           1441.0           1731.0            725.0              0.0   
6229785           1440.0           1731.0            725.0              0.0   
6229786             97.0            231.0             97.0              0.0   
6229787             64.0            242.0             32.0             32.0   
6229788           1459.0           1731.0            741.0              0.0   

         Fwd Pkt Len Mean  ...  Fwd Seg Size Min  Active Mean  Active Std  \
0                0.000000  ...                 0          0.0         0.0   
1               82.600000  ...                32          0.0         0.0   
2               81.642857  ...                32          0.0         0.0   
3               77.437500  ...                20          0.0         0.0   
4               42.800000  ...                32          0.0         0.0   
...                   ...  ...               ...          ...         ...   
6229784        102.928571  ...                20          0.0         0.0   
6229785        102.857143  ...                20          0.0         0.0   
6229786         24.250000  ...                20          0.0         0.0   
6229787         32.000000  ...                 8          0.0         0.0   
6229788        104.214286  ...                20          0.0         0.0   

         Active Max  Active Min   Idle Mean    Idle Std    Idle Max  \
0               0.0         0.0  56320859.5  139.300036  56320958.0   
1               0.0         0.0         0.0    0.000000         0.0   
2               0.0         0.0         0.0    0.000000         0.0   
3               0.0         0.0         0.0    0.000000         0.0   
4               0.0         0.0         0.0    0.000000         0.0   
...             ...         ...         ...         ...         ...   
6229784         0.0         0.0         0.0    0.000000         0.0   
6229785         0.0         0.0         0.0    0.000000         0.0   
6229786         0.0         0.0         0.0    0.000000         0.0   
6229787         0.0         0.0         0.0    0.000000         0.0   
6229788         0.0         0.0         0.0    0.000000         0.0   

           Idle Min   Label  
0        56320761.0  Benign  
1               0.0  Benign  
2               0.0  Benign  
3               0.0  Benign  
4               0.0  Benign  
...             ...     ...  
6229784         0.0  Benign  
6229785         0.0  Benign  
6229786         0.0  Benign  
6229787         0.0  Benign  
6229788         0.0  Benign  

[6229789 rows x 79 columns]

In [3]:
#Examine the distribution of the classes
dist = pd.DataFrame()
dist['Freq'] = events['Label'].value_counts()
dist['Percentage'] = events['Label'].value_counts(normalize=True)
dist.style.format("{:.3%}", subset=['Percentage'])

**Observations:**
* There are 15 classes
* The dataset is highly unbalanced. The "Benign" class has 56% of the samples, while other classes such as "SQL Injection" only has 0.001% of the samples. When we at the training stage, we'll need to look at strategies for making this dataset more balanced.


In [4]:
#Check if any of the columns have all zero data
colZeros = ((events == 0).sum() == len(events))

print('Columns with all zero data:', (colZeros[ colZeros==True ]).count())

if ((colZeros[ colZeros==True ]).count() > 0):
    print('Columns: ', colZeros[ colZeros==True ].index)


Columns with all zero data: 8
Columns:  Index(['Bwd PSH Flags', 'Bwd URG Flags', 'Fwd Byts/b Avg', 'Fwd Pkts/b Avg',
       'Fwd Blk Rate Avg', 'Bwd Byts/b Avg', 'Bwd Pkts/b Avg',
       'Bwd Blk Rate Avg'],
      dtype='object')


In [5]:
#Drop those columns with all zero data - as they will not impact on the solution at all
events.drop(axis=1, columns=colZeros[ colZeros==True ].index, inplace=True)

In [6]:
events.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6229789 entries, 0 to 6229788
Data columns (total 71 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Dst Port           int64  
 1   Protocol           int64  
 2   Flow Duration      int64  
 3   Tot Fwd Pkts       int64  
 4   Tot Bwd Pkts       int64  
 5   TotLen Fwd Pkts    float64
 6   TotLen Bwd Pkts    float64
 7   Fwd Pkt Len Max    float64
 8   Fwd Pkt Len Min    float64
 9   Fwd Pkt Len Mean   float64
 10  Fwd Pkt Len Std    float64
 11  Bwd Pkt Len Max    float64
 12  Bwd Pkt Len Min    float64
 13  Bwd Pkt Len Mean   float64
 14  Bwd Pkt Len Std    float64
 15  Flow Byts/s        float64
 16  Flow Pkts/s        float64
 17  Flow IAT Mean      float64
 18  Flow IAT Std       float64
 19  Flow IAT Max       float64
 20  Flow IAT Min       float64
 21  Fwd IAT Tot        float64
 22  Fwd IAT Mean       float64
 23  Fwd IAT Std        float64
 24  Fwd IAT Max        float64
 25  Fwd IAT Min       

In [7]:
pd.set_option('display.float_format',  '{:,.2f}'.format)
events.describe()

Dst Port     Protocol  Flow Duration  Tot Fwd Pkts  Tot Bwd Pkts  \
count 6,229,789.00 6,229,789.00   6,229,789.00  6,229,789.00  6,229,789.00   
mean     22,644.56         6.50   7,446,591.19         50.25          2.35   
std      25,187.99         2.35  22,580,670.44      2,453.24         69.71   
min           0.00         0.00           0.00          1.00          0.00   
25%          80.00         6.00          52.00          2.00          0.00   
50%       3,389.00         6.00       4,221.00          2.00          1.00   
75%      51,073.00         6.00   1,315,446.00          3.00          2.00   
max      65,535.00        17.00 120,000,000.00    309,629.00     69,241.00   

       TotLen Fwd Pkts  TotLen Bwd Pkts  Fwd Pkt Len Max  Fwd Pkt Len Min  \
count     6,229,789.00     6,229,789.00     6,229,789.00     6,229,789.00   
mean          1,804.40           906.64           183.17             2.53   
std          99,436.04        94,460.73           333.70            19.94   
min               0.00             0.00             0.00             0.00   
25%               0.00             0.00             0.00             0.00   
50%               0.00             0.00             0.00             0.00   
75%             259.00           296.00           161.00             0.00   
max     144,391,846.00   101,000,000.00        34,368.00         1,460.00   

       Fwd Pkt Len Mean  ...  Fwd Act Data Pkts  Fwd Seg Size Min  \
count      6,229,789.00  ...       6,229,789.00      6,229,789.00   
mean              41.66  ...              47.81             22.78   
std               72.83  ...           2,453.20              7.01   
min                0.00  ...               0.00              0.00   
25%                0.00  ...               0.00             20.00   
50%                0.00  ...               0.00             20.00   
75%               53.33  ...               1.00             24.00   
max            7,303.31  ...         309,628.00             48.00   

         Active Mean    Active Std     Active Max     Active Min  \
count   6,229,789.00  6,229,789.00   6,229,789.00   6,229,789.00   
mean       40,067.23     14,685.77      55,804.13      30,376.80   
std       967,185.82    554,115.04   1,267,571.60     827,119.04   
min             0.00          0.00           0.00           0.00   
25%             0.00          0.00           0.00           0.00   
50%             0.00          0.00           0.00           0.00   
75%             0.00          0.00           0.00           0.00   
max   114,000,000.00 72,200,000.00 114,000,000.00 114,000,000.00   

           Idle Mean      Idle Std       Idle Max       Idle Min  
count   6,229,789.00  6,229,789.00   6,229,789.00   6,229,789.00  
mean    5,517,036.16     59,057.82   5,581,183.88   5,464,611.79  
std    19,028,510.87  1,077,565.87  19,134,593.28  18,996,417.90  
min             0.00          0.00           0.00           0.00  
25%             0.00          0.00           0.00           0.00  
50%             0.00          0.00           0.00           0.00  
75%             0.00          0.00           0.00           0.00  
max   120,000,000.00 76,393,949.17 120,000,000.00 120,000,000.00  

[8 rows x 70 columns]

### Data quality checks


In [9]:
#Check if any of the columns have any missing/NaN data
colNulls = events.isna().sum()
print('Columns with NaN values:', (colNulls[ colNulls != 0 ]).count())
if ((colNulls[ colNulls != 0 ]).count() > 0):
    print('Columns: ', colNulls[ colNulls != 0 ])


Columns with NaN values: 1
Columns:  Flow Byts/s    59215
dtype: int64


In [10]:
#What is the value of the Flow Pkts/s feature (which is similar to the Flow Byts/s feature)
events[ events['Flow Byts/s'].isnull()]['Flow Pkts/s'].value_counts()

Flow Pkts/s
inf    59215
Name: count, dtype: int64

In [11]:
#What is the value of the Flow Duration feature (as Flow Byts/s metric is calculated using the Flow Duration feature - eg Flow Bytes/s = Bytes in Flow / Flow Duration)
events[ events['Flow Byts/s'].isnull()]['Flow Duration'].value_counts()

Flow Duration
0    59215
Name: count, dtype: int64

In [12]:
#How many samples are affected by this Flow Duration=0 issue?
events[ events['Flow Duration']==0]['Flow Byts/s'].value_counts()

Flow Byts/s
inf    34621
Name: count, dtype: int64

In [13]:
#How many samples are affected by this Flow Duration=0 issue?
events[ events['Flow Duration']==0]['Label'].value_counts()

Label
Benign            92535
Infilteration      1295
FTP-BruteForce        6
Name: count, dtype: int64

**Observation:** The above analysis shows that the root cause of the 59215 samples having NaN in the 'Flow Pkts/s' features is due to the Flow Duration feature being set to 0 for whatever reason (which then in turn causes division-by-zero error when other features are being calculated). Given the Flow Duration feature is measured in microseconds, we expect every sample should have a valid value here. Some error would have occured during the network packet capture or during the generation of flow related features.
As this issue has also resulted in NaN and INF values in the 'Flow Byts/s' and 'Flow Pkts/s' features, it would make sense to completely remove these samples altogether (even before we get to the stage of allocating training vs testing sets). The impact of the removal would also be small as there are only 93836 samples afflicted by this issue

In [14]:
events.drop(axis=1, index=events[ events['Flow Duration']==0].index, inplace=True)

In [15]:
events.count()

Dst Port         6135953
Protocol         6135953
Flow Duration    6135953
Tot Fwd Pkts     6135953
Tot Bwd Pkts     6135953
                  ...   
Idle Mean        6135953
Idle Std         6135953
Idle Max         6135953
Idle Min         6135953
Label            6135953
Length: 71, dtype: int64

In [16]:
#Check for duplicates in the dataset
dupRows = events.duplicated()
print('Duplicate Rows: ', dupRows.sum(), 'Duplicated %:', dupRows.sum()/len(events)*100)


Duplicate Rows:  1542894 Duplicated %: 25.145140453324856


**Observations:** 25% of the Dataset seems to have duplicated data. This is expected for some of the intrusion attempts as they carry out repeated attacks and features such as Timestamp, IP addressess...etc have been removed from the dataset. At the training stage, we will look at deleting some of the duplicated samples in order to reduce the sample sizes to aid training.

### Delete invalid data

During the evalution stage of a suitable classifier, it became apparent from analysis on the Confusion Matrix, that some of the 'FTP-BruteForce' samples were being incorrectly predicted as 'DoS attacks-SlowHTTPTest', while large number of the 'DoS attacks-SlowHTTPTest' samples were being incorrectly predicted as 'FTP-BruteForce'. Further inspection of these samples revealed that they are almost identical. 

In [22]:
from CustomFunctions import printLarge

printLarge(events[ events['Label']=='DoS attacks-SlowHTTPTest'])

Dst Port  Protocol  Flow Duration  Tot Fwd Pkts  Tot Bwd Pkts  \
876639         21         6             21             1             1   
876640         21         6              3             1             1   
876641         21         6              3             1             1   
876642         21         6              3             1             1   
876643         21         6              2             1             1   
...           ...       ...            ...           ...           ...   
1925111        21         6              3             1             1   
1925112        21         6              3             1             1   
1925113        21         6              3             1             1   
1925114        21         6              5             1             1   
1925115        21         6              2             1             1   

         TotLen Fwd Pkts  TotLen Bwd Pkts  Fwd Pkt Len Max  Fwd Pkt Len Min  \
876639              0.00             0.00             0.00             0.00   
876640              0.00             0.00             0.00             0.00   
876641              0.00             0.00             0.00             0.00   
876642              0.00             0.00             0.00             0.00   
876643              0.00             0.00             0.00             0.00   
...                  ...              ...              ...              ...   
1925111             0.00             0.00             0.00             0.00   
1925112             0.00             0.00             0.00             0.00   
1925113             0.00             0.00             0.00             0.00   
1925114             0.00             0.00             0.00             0.00   
1925115             0.00             0.00             0.00             0.00   

         Fwd Pkt Len Mean  Fwd Pkt Len Std  Bwd Pkt Len Max  Bwd Pkt Len Min  \
876639               0.00             0.00             0.00             0.00   
876640               0.00             0.00             0.00             0.00   
876641               0.00             0.00             0.00             0.00   
876642               0.00             0.00             0.00             0.00   
876643               0.00             0.00             0.00             0.00   
...                   ...              ...              ...              ...   
1925111              0.00             0.00             0.00             0.00   
1925112              0.00             0.00             0.00             0.00   
1925113              0.00             0.00             0.00             0.00   
1925114              0.00             0.00             0.00             0.00   
1925115              0.00             0.00             0.00             0.00   

         Bwd Pkt Len Mean  Bwd Pkt Len Std  Flow Byts/s  Flow Pkts/s  \
876639               0.00             0.00         0.00    95,238.10   
876640               0.00             0.00         0.00   666,666.67   
876641               0.00             0.00         0.00   666,666.67   
876642               0.00             0.00         0.00   666,666.67   
876643               0.00             0.00         0.00 1,000,000.00   
...                   ...              ...          ...          ...   
1925111              0.00             0.00         0.00   666,666.67   
1925112              0.00             0.00         0.00   666,666.67   
1925113              0.00             0.00         0.00   666,666.67   
1925114              0.00             0.00         0.00   400,000.00   
1925115              0.00             0.00         0.00 1,000,000.00   

         Flow IAT Mean  Flow IAT Std  Flow IAT Max  Flow IAT Min  Fwd IAT Tot  \
876639           21.00          0.00         21.00         21.00         0.00   
876640            3.00          0.00          3.00          3.00         0.00   
876641            3.00          0.00          3.00          3.00         0.00   
876642            3.0

In [21]:
from CustomFunctions import printLarge

printLarge(events[ events['Label']=='FTP-BruteForce'])

Dst Port  Protocol  Flow Duration  Tot Fwd Pkts  Tot Bwd Pkts  \
33            21         6             19             1             1   
34            21         6              3             1             1   
35            21         6              3             1             1   
36            21         6              2             1             1   
37            21         6              2             1             1   
...          ...       ...            ...           ...           ...   
382711        21         6              1             1             1   
382712        21         6              2             1             1   
382715        21         6              1             1             1   
382716        21         6              1             1             1   
382719        21         6              1             1             1   

        TotLen Fwd Pkts  TotLen Bwd Pkts  Fwd Pkt Len Max  Fwd Pkt Len Min  \
33                 0.00             0.00             0.00             0.00   
34                 0.00             0.00             0.00             0.00   
35                 0.00             0.00             0.00             0.00   
36                 0.00             0.00             0.00             0.00   
37                 0.00             0.00             0.00             0.00   
...                 ...              ...              ...              ...   
382711             0.00             0.00             0.00             0.00   
382712             0.00             0.00             0.00             0.00   
382715             0.00             0.00             0.00             0.00   
382716             0.00             0.00             0.00             0.00   
382719             0.00             0.00             0.00             0.00   

        Fwd Pkt Len Mean  Fwd Pkt Len Std  Bwd Pkt Len Max  Bwd Pkt Len Min  \
33                  0.00             0.00             0.00             0.00   
34                  0.00             0.00             0.00             0.00   
35                  0.00             0.00             0.00             0.00   
36                  0.00             0.00             0.00             0.00   
37                  0.00             0.00             0.00             0.00   
...                  ...              ...              ...              ...   
382711              0.00             0.00             0.00             0.00   
382712              0.00             0.00             0.00             0.00   
382715              0.00             0.00             0.00             0.00   
382716              0.00             0.00             0.00             0.00   
382719              0.00             0.00             0.00             0.00   

        Bwd Pkt Len Mean  Bwd Pkt Len Std  Flow Byts/s  Flow Pkts/s  \
33                  0.00             0.00         0.00   105,263.16   
34                  0.00             0.00         0.00   666,666.67   
35                  0.00             0.00         0.00   666,666.67   
36                  0.00             0.00         0.00 1,000,000.00   
37                  0.00             0.00         0.00 1,000,000.00   
...                  ...              ...          ...          ...   
382711              0.00             0.00         0.00 2,000,000.00   
382712              0.00             0.00         0.00 1,000,000.00   
382715              0.00             0.00         0.00 2,000,000.00   
382716              0.00             0.00         0.00 2,000,000.00   
382719              0.00             0.00         0.00 2,000,000.00   

        Flow IAT Mean  Flow IAT Std  Flow IAT Max  Flow IAT Min  Fwd IAT Tot  \
33              19.00          0.00         19.00         19.00         0.00   
34               3.00          0.00          3.00          3.00         0.00   
35               3.00          0.00          3.00          3.00         0.00   
36               2.00          0.00          2.00          2.00         

**Observation:** Note how nearly all the features look almost identical. Even these highly variable features show similarities: [Init Fwd Win Bytes], [Fwd Seg Size Min], [Fwd Pkts/s], [Fwd Header Len]. What is also surprising is that the DST Port feature is "21" for both of these classes. Port 21 is expected for FTP-BruteForce attacks (as port 21 is a well known port for FTP). However, for 'DoS attacks-SlowHTTPTest', the expected port is either 80 (for HTTP attacks) or port 443 (for HTTPS attacks). So it is highly likely that the 'DoS attacks-SlowHTTPTest' samples were created with the incorrect label. Most likely those samples are actually FTP-Bruteforce attacks, but incorrectly labelled as 'DoS attacks-SlowHTTPTest'.

Hence the decision was made to completely delete these 'DoS attacks-SlowHTTPTest' samples.

In [ ]:
print('Dataset size before deleting DoS attacks-SlowHTTPTest samples: ', events.shape)
print('DoS attacks-SlowHTTPTest samples to be deleted: ', len(events[ events['Label']=='DoS attacks-SlowHTTPTest']))
events.drop(axis=0, index=events[ events['Label']=='DoS attacks-SlowHTTPTest'].index, inplace=True)
print('Dataset size after deleting DoS attacks-SlowHTTPTest samples: ', events.shape)

### Output final dataset

In [34]:
#Output all data to a single "cleaned" file that is ready for training at the next stage
filename_out = 'Inputs\\cleaned\\DataReady.csv'
events.to_csv(filename_out, index=False)